Web5 cli guide (#1568)
TBD54566975/developer.tbd.website: site/docs/web5/cli.mdx

---
sidebar_position: 6
title: CLI
hide_title: true
hide_table_of_contents: false
---

# Web5 CLI

The Web5 CLI is a Rust-based command-line tool for creating and managing [Decentralized Identifiers (DIDs)](/docs/web5/decentralized-identifiers/what-are-dids) and [Verifiable Credentials (VCs)](/docs/web5/verifiable-credentials/what-are-vcs).

This guide will walk you through installation and usage of the Web5 CLI.

## Installation

To install the Web5 CLI, run:

```bash
curl -L https://raw.githubusercontent.com/TBD54566975/web5-rs/main/crates/web5_cli/install.sh | bash -s v0.0.4
```

## Creating a DID

> The `web5 did create` command [creates a DID](/docs/web5/decentralized-identifiers/how-to-create-did) using different DID methods such as `did:dht`, `did:jwk`, and `did:web`. 

Reviewer: are these 3 the only methods?

maybe they should be listed out like `Options` are and remove them from this sentence. You should link them to the the glossary entries as well so people can compare their method choices.

Author: okay done. so far only three did methods are supported

> The `web5 did create` command [creates a DID](/docs/web5/decentralized-identifiers/how-to-create-did) using different DID methods such as `did:dht`, `did:jwk`, and `did:web`. 

Reviewer: dont think you should link to the Create a DID guide since that gives them the SDK approach. I imagine that would be really confusing. Maybe instead, in the intro you can say "we support creating a DID via [SDK](link here) or via CLI, which is detailed in this guide"

Author: good point

```bash
web5 did create <method> [options]
```

### Options
* `--no-indent`: Output the DID document without indentation.
* `--json-escape`: Escape JSON strings in the output.
* `--domain <domain>`: For did:web, specify the domain for the DID Web creation.

> * `--no-publish`: For the dht method, prevents publishing the DID immediately.

Reviewer: is JWK automatically published, or not?

Author: did:jwk's can't get published -- see Gabe's comment on Discord here: https://discord.com/channels/937858703112155166/1245544729345527829/1245787004713701478

The no-publish flag is only for dht. Should I write anything to make it more clear?

### Examples

```bash
# Create a did:jwk with no indent and json escape
web5 did create jwk --no-indent --json-escape

# Create a did:web 
web5 did create web --domain https://example.com

# Create a did:dht without publishing
web5 did create dht --no-publish
```

## Resolving a DID

> The `web5 did resolve` command [resolves a DID](/docs/web5/decentralized-identifiers/how-to-resolve-a-did).

Reviewer: ```suggestion
The `web5 did resolve` command resolves a DID to retrieve its [DID Document](/docs/web5/decentralized-identifiers/did_documents)
```

```bash
web5 did resolve <did-uri> [options]
```

> ### Options

Reviewer: can you bold all the Options and Examples headings (instead of ###) so they dont clutter the right sidebar?

* `--no-indent`: Output the DID document without indentation.
* `--json-escape`: Escape JSON strings in the output.

### Examples

```bash
web5 did resolve did:dht:bwrh9fphhegh1jn3hhe656zuyfxqdm493so5r4hmpte1yf456w5o
```

## Create and Sign a VC

> The `web5 vc create` command [creates and sign VCs](/docs/web5/verifiable-credentials/vc-issuance).

Reviewer: ```suggestion
The `web5 vc create` command creates and sign verifiable credentials. It returns the VC in [JWT](https://jwt.io/) format.
```

```bash
web5 vc create <credential-subject-id> [options]
```

### Options

> * `--portable-did <did>`: specifies the DID used to sign the Verifiable Credential. If omitted, VC will be unsigned

Reviewer: is this different than issuer DID? trying to understand why i would need to provide both

> * `--issuer <issuer-did>`: Specifies the issuer of the credential. If issuer is provided issuer without `--portable-did`, the VC will be unsigned.

Reviewer: ```suggestion
* `--issuer <issuer-did>`: Specifies the issuer of the credential. If issuer is provided without `--portable-did`, the VC will be unsigned.
```

> * `--expiration-date <date>`: Set an expiration date for the credential.

Reviewer: in what format?

* `--no-indent`: Output the VC without indentation.
* `--json-escape`: Escape JSON strings in the output.

### Examples
```bash
## Portable DID that will be used as the issuer and to sign the credential
export PORTABLE_DID=$(web5 did create dht --no-indent)

## Create and sign VC with portable DID

> web5 vc create "alice" --portable-did $PORTABLE_DID

Reviewer: instead of "alice", show a real DID.

## Create and sign VC with an expiration date

> web5 vc create "alice" --portable-did $PORTABLE_DID --expiration-date 2025-12-31

Reviewer: instead of "alice", show a real DID.

Author: done

> ```

Reviewer: add one more example for unsigned vc

```
## Create unsigned VC
web5 vc create "real_did_here" --issuer "real_did_here"
```
can issuer DID's ID be obtained from the portable did?

## Verify a VC

> The `web5 vc verify` command [verifies a VC](/docs/web5/verifiable-credentials/verify-vc) in JWT format.

Reviewer: ```suggestion
The `web5 vc verify` command verifies a VC that is in JWT format.
```

Author: done

```bash
web5 did create <method> [options]
```

**Supported DID methods**
- `did:dht`: [Distributed Hash Table](/docs/glossary#did-method---dht)
- `did:jwk`: [JSON Web Key](/docs/glossary#did-method---jwk)
- `did:web`: [Web](/docs/glossary#did-method---web)

**Options**
* `--no-indent`: Output the DID document without indentation.
* `--json-escape`: Escape JSON strings in the output.
* `--domain <domain>`: For did:web, specify the domain for the DID Web creation.
* `--no-publish`: Prevents immediately publishing for `did:dht`.

**Examples**

```bash
# Create a did:dht without publishing
web5 did create dht --no-publish

# Create a did:jwk with no indent and json escape
web5 did create jwk --no-indent --json-escape

# Create a did:web 
web5 did create web --domain https://example.com
```

## Resolve a DID

The `web5 did resolve` command resolves a DID to retrieve its [DID Document](/docs/web5/decentralized-identifiers/did_documents)

```bash
web5 did resolve <did-uri> [options]
```

**Options**
* `--no-indent`: Output the DID document without indentation.
* `--json-escape`: Escape JSON strings in the output.

> **Examples**

Reviewer: ```suggestion
**Example**
```

```bash
web5 did resolve did:dht:bwrh9fphhegh1jn3hhe656zuyfxqdm493so5r4hmpte1yf456w5o
```

> ## Create an Unsigned VC

Reviewer: hmm does this need its own section? can we just do "Create a VC" and add one example of creating an unsigned VC? btw, what does that return, VC as JSON?

Author: OHHHH i misunderstood your initial comment about adding info about Creating an unsigned VC

The `web5 vc create` command can create a Verifiable Credential without signing it.

```bash
web5 vc create <credential-subject-did-uri> --issuer <issuer-did-uri> [options]
```
**Options**
* `--issuer <issuer-did-uri>`: Specifies the issuer of the credential.

> * `--expiration-date <date>`: Set an expiration date for the credential.

Reviewer: ```suggestion
* `--expiration-date <date>`: Set an expiration date for the credential in YYYY-MM-DD format
```

```

**Options**
* `--portable-did <did>`: Specifies the DID used to sign the Verifiable Credential and outputs it as a JWT. If included, the Portable DID's URI is automatically set as the issuer.
* `--issuer <issuer-did-uri>`: Specifies the issuer of the credential.
* `--expiration-date <date>`: Set an expiration date for the credential.
* `--no-indent`: Output the VC without indentation.
* `--json-escape`: Escape JSON strings in the output.

**Examples**
```bash
# Portable DID that will be used as the issuer and to sign the credential
export PORTABLE_DID=$(web5 did create dht --no-indent)

# Create and sign VC with portable DID
web5 vc create "did:dht:cm1yqfjzfdtauh33nauwf3sqsijqziwrydicr8dbtho3cucb9nno" --portable-did $PORTABLE_DID

# Create and sign VC with an expiration date
web5 vc create "did:dht:cm1yqfjzfdtauh33nauwf3sqsijqziwrydicr8dbtho3cucb9nno" --portable-did $PORTABLE_DID --expiration-date 2025-12-31
```

## Verify a VC
The `web5 vc verify` command verifies a VC that is in JWT format.

```bash
web5 vc verify <vc-jwt> [options]
```

**Options**
* `--no-indent`: Outputs the verification result without indentation.
* `--json-escape`: Escapes JSON strings in the output.

> **Examples**

Reviewer: ```suggestion
**Example**
```
